File: binded/kt/src/main/kotlin/web5/sdk/dids/did/methods/jwk/DidJwk.kt
package web5.sdk.dids.did.methods.jwk

import web5.sdk.crypto.keys.Jwk
import web5.sdk.dids.did.methods.Did
import web5.sdk.dids.did.methods.DidDocument
import web5.sdk.dids.did.methods.DidResolutionResult

import web5.sdk.DidJwk as RcbDidJwk

class DidJwk {
    val did: Did
    val document: DidDocument

    constructor(publicKey: Jwk) {
        val didJwk = RcbDidJwk.fromPublicKey(publicKey.toBinding());

        this.did = Did.fromBinding(didJwk.getData().did)
        this.document = DidDocument.fromBinding(didJwk.getData().document)
    }

    constructor(uri: String) {
        val resolutionResult = resolve(uri)
        val verificationMethod = resolutionResult.didDocument?.verificationMethod?.get(0)
        val publicKey = verificationMethod?.publicKeyJwk!!

        val rcbDidJwk = RcbDidJwk.fromPublicKey(publicKey.toBinding());

        this.did = Did.fromBinding(rcbDidJwk.getData().did)
        this.document = DidDocument.fromBinding(rcbDidJwk.getData().document)
    }

    companion object {
        @JvmStatic
        fun resolve(uri: String): DidResolutionResult {
            return TODO("Provide the return value")
        }
    }
}

File: bind/uniffi/src/dids/methods/did_jwk.rs
use crate::errors::RcbResult;
use std::sync::Arc;
use web5::apid::{dids::methods::did_jwk::DidJwk, jwk::Jwk};

pub struct RcbDidJwk(DidJwk);

impl RcbDidJwk {
    pub fn from_public_jwk(public_key: Jwk) -> RcbResult<Self> {
        let did_jwk = DidJwk::from_public_jwk(public_key).map_err(|e| Arc::new(e.into()))?;
        Ok(Self(did_jwk))
    }

    pub fn from_uri(uri: &str) -> RcbResult<Self> {
        let did_jwk = DidJwk::from_uri(uri).map_err(|e| Arc::new(e.into()))?;
        Ok(Self(did_jwk))
    }

    // 🚧
    // pub fn resolve(_uri: &str) -> ResolutionResult {
    //
    // }

    pub fn get_data(&self) -> DidJwk {
        self.0.clone()
    }
}
